src/crypto/key_exchange.h

```diff
@@ -4,199 +4,118 @@
+      std::shared_ptr<crypto::KeyPair> own_kp,
+      std::shared_ptr<crypto::PublicKey> peer_pubk) :
+      curve(own_kp->get_curve_id())
     {
-      auto tmp_ctx =
-        crypto::mbedtls::make_unique<crypto::mbedtls::ECDHContext>();
-
-      int rc = mbedtls_ecdh_get_params(
-        tmp_ctx.get(),
-        mbedtls_pk_ec(*own_kp->get_raw_context()),
-        MBEDTLS_ECDH_OURS);
-      if (rc != 0)
-      {
-        throw std::logic_error(crypto::error_string(rc));
-      }
-
-      rc = mbedtls_ecdh_get_params(
-        tmp_ctx.get(),
-        mbedtls_pk_ec(*peer_pubk->get_raw_context()),
-        MBEDTLS_ECDH_THEIRS);
-      if (rc != 0)
-      {
-        throw std::logic_error(crypto::error_string(rc));
-      }
-      ctx = std::move(tmp_ctx);
+      own_key = own_kp;
+      peer_key = peer_pubk;
     }
 
-    void free_ctx()
-    {
-      // Should only be called when shared secret has been computed.
-      ctx.reset();
-    }
+    ~KeyExchangeContext() {}
 
-    ~KeyExchangeContext()
+    std::vector<uint8_t> get_own_key_share() const
     {
-      free_ctx();
+      if (!own_key)
+      {
+        throw std::runtime_error("missing node key");
+      }
 
-      OPENSSL_cleanse(key_share.data(), key_share.size());
+      // For backwards compatability we need to keep the format we used with
+      // mbedTLS, which is the raw EC point with an extra size byte in the
+      // front.
+      auto tmp = own_key->public_key_raw();
+      tmp.insert(tmp.begin(), tmp.size());
+      return tmp;
     }
 
-    const std::vector<uint8_t>& get_own_key_share() const
+    std::vector<uint8_t> get_peer_key_share() const
     {
-      if (!ctx)
-      {
-        throw std::runtime_error("Missing key exchange context");
-      }
-
-      if (key_share.empty())
+      if (!peer_key)
       {
-        throw std::runtime_error("Missing node key share");
+        throw std::runtime_error("missing peer key");
       }
 
-      // Note that this function returns a vector of bytes
-      // where the first byte represents the
-      // size of the public key
-      return key_share;
-    }
-
-    const std::vector<uint8_t>& get_peer_key_share() const
-    {
-      return peer_key_share;
+      auto tmp = peer_key->public_key_raw();
+      tmp.insert(tmp.begin(), tmp.size());
+      return tmp;
     }
 
     void reset()
     {
-      key_share.clear();
-      peer_key_share.clear();
-      ctx.reset();
-      create_fresh_key_share();
+      peer_key.reset();
+      own_key = make_key_pair(crypto::CurveID::SECP384R1);
     }
 
     void load_peer_key_share(const std::vector<uint8_t>& ks)
     {
-      load_peer_key_share({ks.data(), ks.size()});
-    }
-
-    void load_peer_key_share(CBuffer ks)
-    {
-      if (!ctx)
+      if (ks.size() == 0)
       {
-        throw std::runtime_error(
-          "Missing key exchange context when loading peer key share");
+        throw std::runtime_error("Missing peer key share");
```

Comment:
I don't care either way, but we shouldn't have half our exceptions descriptions capitalised and the other half not, it looks sloppy.

Reply:
I completely agree, it seems to be totally random throughout the codebase though. I'll fix my brain to stick to lower-case for now.